package Proto;

message Stroke {
	message Point {
		required float x = 1;
		required float y = 2;
		optional float t = 3;
	}
	repeated Point points = 1;
	optional uint32 button = 2 [default = 0];
	optional uint32 trigger = 3 [default = 0];
	optional bool timeout = 4 [default = false];
}

message Action {
	enum Type {
		Command = 0;
		SendKey = 1;
		Scroll = 2;
		Ignore = 3;
		Button = 4;
		Unminimize = 5;
		ShowHide = 6;
		Disable = 7;
	}
	optional Type type = 1;
	optional string cmd = 2;
	optional int32 button = 3;
	optional uint32 mods = 4;
}

message ActionList {
	message StrokeInfo {
		required uint32 id = 1;
		repeated Stroke strokes = 2;
		optional Action action = 3;
		optional string name = 4;
	}

	optional string name = 1;
	repeated uint32 deleted = 2;
	repeated StrokeInfo added = 3;
	repeated ActionList children = 4;
	optional bool app = 5 [default = true];
}

message Button {
	required int32 button = 1;
	optional uint32 modifiers = 2 [default = 0];
}

message Prefs {
	message Exception {
		required string name = 1;
		optional Button button = 2;
	}

	enum Trace {
		Default = 0;
		Shape = 1;
		NoTrace = 2;
		Annotate = 3;
		Fire = 4;
		Water = 5;
	}

	enum Timeout {
		Off = 0;
		Conservative = 1;
		Medium = 2;
		Aggressive = 3;
		Flick = 4;
		Custom = 5;
	}

	message Color {
		required int32 red = 1;
		required int32 green = 2;
		required int32 blue = 3;
		optional int32 alpha = 4 [default = 65535];
	}

	repeated Exception exceptions = 1;
	optional float p = 2 [default = 0.5];
	optional Button button = 3;
	optional Trace trace = 4 [default = Default];
	optional bool advanced_ignore = 5 [default = false];
	optional int32 radius = 6 [default = 16];
	optional bool ignore_grab = 7 [default = false];
	optional bool timing_workaround = 8 [default = false];
	optional bool pressure_abort = 9 [default = false];
	optional int32 pressure_threshold = 10 [default = 192];
	optional bool proximity = 11 [default = false];
	optional bool popups = 12 [default = true];
	optional bool left_handed = 13 [default = false];
	optional int32 init_timeout = 14 [default = 200];
	optional int32 min_speed = 15 [default = 40];
	optional Timeout timeout_profile = 16 [default = Conservative];
	optional bool timeout_gestures = 17 [default = false];
	optional bool tray_icon = 18 [default = true];
	repeated string excluded_devices = 19;
	optional Color color = 20;
	optional int32 trace_width = 21 [default = 5];
	repeated Button extra_buttons = 22;
	optional bool advanced_popups = 23 [default = true];
}
